README.md
# Switchable

---

[![spm package](http://spmjs.io/badge/arale-switchable)](http://spmjs.io/package/arale-switchable)
[![Build Status](https://secure.travis-ci.org/aralejs/switchable.png)](https://travis-ci.org/aralejs/switchable)
[![Coverage Status](https://coveralls.io/repos/aralejs/switchable/badge.png?branch=master)](https://coveralls.io/r/aralejs/switchable)


Switchable 提供了切换的基本操作，并通过参数配置实现了自动播放，循环，切换等操作，并提供不同的特点封装成了Tabs, Slide, Carousel, Accordion模块。用户可以根据自己的需求引用不同的模块。

---

Switchable 继承了 [widget](http://aralejs.org/widget/)，可使用其中包括 [base](http://aralejs.org/base/)、[class](http://aralejs.org/class/)、[events](http://aralejs.org/events/)、[attribute](http://aralejs.org/base/docs/attribute.html)、[aspect](http://aralejs.org/base/docs/aspect.html) 的属性和方法。

## 配置说明

### triggers `String|Array`
    
触发器列表, 支持直接传入选择器，也可以是元素数组。
   
### panels `String|Array`

面板列表，支持直接传入选择器，也可以是元素数组。

**注意: 如果页面上有多个 Switchable 组件, triggers 和 panels 的 selector 最好写的更加精确些, 比如 `triggers: '#banner .ui-switchable-nav li'`,
而不是 `triggers: '.ui-switchable-nav li'`, 这样能避免多个组件引用 triggers/panels 的冲突.**

### triggerType `String`
    
触发类型，默认`hover`.还可以选择`click`
 
### delay `Number`

默认为`100`， 触发器延迟时间，单位为毫秒。
    
### activeTriggerClass `String`
    
触发器被选中时的class,默认`ui-switchable-active`。

### activeIndex `Number`

初始化时，自动切换到指定面板，默认为`0`，也就是第一个。

### step `Number`

步长，表示每次切换时需要间隔多少个panels, 默认为`1`。

### viewSize `Array`

可见视图区域的大小. 如果 css 中不设置 panel 的高宽或初始没有 panel , 则需要这里手工指定大小, 默认为 [].
  
`当 panel 的高宽 css 中不指定时，需要设置 viewSize 为单个 panel 的高宽.`
  
### autoplay `Boolean`

是否自动切换，默认为`false`, 开启后，不需要触发触发器，可以实现自动播放。

### interval `Number`

自动播放间隔时间, 以毫秒为单位, 默认为 `3000`。

### circular `Boolean`

是否循环切换, 默认为 `true`, 是否循环播放, 当切换到最初/最后一个时, 是否切换到最后/最初一个。
 
### effect `String`

动画效果函数, 默认没有特效, 可取 `scrollx`, `scrolly`, `fade` 或者直接传入自定义效果函数.

### duration `Number`

默认为 `500`, 以毫秒为单位， 动画的时长.

### easing `String|Function`

动画效果，目前支持,`easeNode`默认, `easeIn`, `easeOut`, `easeBoth`, `easeInStrong`, `easeOutStrong`, `easeBothStrong`, `elasticIn`, `elasticOut`, `elasticBoth`, `backIn`, `backOut`, `backBoth`, `bounceIn`, `bounceOut`, `bounceBoth`。


## DATA API 配置属性

目前项目支持data-api的形式配置相关属性，并以role的形式配置到项目中，主要有下面6个role.
前面4个是所有组件均有效，后面2个，只在 Carousel 组件中生效.

其中前面4个仅在在属性配置中没有发现对应的`panels` 和 `triggers` 相关配置时，才会生效。
并且如果同时配置了下面的role， panel的优先级高于content. trigger高于nav.
 
* `panel` 单个面板

* `content` 面板列表元素

* `trigger` 触发器

* `nav` 触发器容器

```html
<!-- 容器元素 -->
<div id="J_Slide">
    <span data-role="prev">上一页</span>
    <ul class="ui-switchable-nav" data-role="nav">
        <li data-role="trigger">标题 A</li>
        <li data-role="trigger">标题 B</li>
        <li data-role="trigger">标题 C</li>
        <li data-role="trigger">标题 D</li>
        <li data-role="trigger">标题 E</li>
    </ul>
    <div class="ui-switchable-content" data-role="content">  <!-- 面板列表 -->
        <div data-role="panel">内容 A</div>
        <div style="display: none" data-role="panel">内容 B</div>
        <div style="display: none" data-role="panel">内容 C</div>
        <div style="display: none" data-role="panel">内容 D</div>
    </div>
    <span data-role="next">下一页</span>
</div>
```
* `prev` 上一页 (仅在Carousel组件生效)

* `next` 下一页 (仅在Carousle组件中生效)

---

## 具体组件配置属性

有些属性在对应类型的模块可能有不同的默认值，还有一些特有的配置属性。

## Accordion (手风琴)

### triggerType `String`

在此模块中，默认值为`click`

### multiple `Boolean`
  
是否支持多个面板展开，默认为`false`。


## Carousel (旋转木马)

### circular `Boolean`

是否支持循环切换。默认`true`

### prevBtn `selector`
    
指定 `前一个`触发器. 默认会加上类名`ui-switchable-prev-btn`。

### nextBtn `selector`
  
指定 `后一个`触发器. 默认会加上类名 `ui-switchable-next-btn`。

### disabledBtnClass `String`

prev/next 按钮不可用状态时的样式类

## Slide (卡盘)

### autoplay `Boolean`

  是否自动切换，默认为`true`。

### circular `Boolean`

  是否循环切换，默认为`true`。


## Tabs (普通标签页)

配置和基础类相同。

### element `HTMLElement`

  容器元素。

### obj.get('triggers') `Array`

  触发器集合， 可能为空。

### obj.get('panels') `Array`

  切换面板结合，可以为空值。

### content `HTMLElement`

  存放面板的容器元素。

### length `Number`

  只读, 触发器或面板的个数。

### .switchTo(toIndex, fromIndex)

  切换到某个视图。
      toIndex {Number* 要切换到的项。
      fromInex {Number* 当前项.
      
### .prev()

  切换到上一视图。

### .next()

  切换到下一视图。

### .stop()

  停止自动切换。只有设置了 autoplay true 时有效。

### .start()

  开始自动切换。

### .destroy()

  组件销毁。

## 组件触发事件

在组件运行中，会触发相关事件，使用者可以根据自己的需要去监听对应的事件

### switch
  
  面板切换前触发。

  ```js
  instance.on('switch', function(toIndex, fromIndex) {
  });
  ```

### switched

  面板切换后触发。

  ```js
  instance.on('switched', function(toIndex, fromIndex) {
  });
  ```

### change:activeIndex

  面板改变时触发。

  ```js
  instance.on('change:activeIndex', function(toIndex, fromIndex) {
  });
  ```


## 最佳实践

1. 直接使用：

    ```js
    seajs.use(['tabs'], function(Tabs) {
        var t = new Tabs({
            element: '#demo1',
            switchTo: 1,
            effect: 'fade'
        });
    });
    ```

2. 也可以适用自动渲染。详情可以参考[examples/autorender.html](../examples/autorender.html)




HISTORY.md
# History

---

## 1.1.1

`tag:fixed` 修复文档，和 main 文件路径。

## 1.1.0

`tag:changed` 迁移 spm@3.x

## 1.0.3

`tag:changed` [#40](https://github.com/aralejs/switchable/issues/40) data-switchable-role 应该换成 data-role 暂时先改回来.


## 1.0.2

`tag:improved` 有些地方 `$(selector, element)` 改成 `this.$(selector)`, 查找当前 element 下的元素

`tag:changed` [#32](https://github.com/aralejs/switchable/issues/32) data-role 应该换成 data-switchable-role

`tag:fixed` [#33](https://github.com/aralejs/switchable/issues/33) Effect width max value in 360.

## 1.0.1

`tag:fixed` [#29](https://github.com/aralejs/switchable/issues/29) 循环切换首个到末个时, 无须经过中间 panel , 而应直接切换.

`tag:fixed` [#28](https://github.com/aralejs/switchable/issues/28) Effect width max value.

## 1.0.0

`tag:fixed` [#20](https://github.com/aralejs/switchable/issues/20) Circular bugfix.

`tag:changed` 去除配置项 pauseOnHover, pauseOnScroll.

`tag:changed` 去除 obj.triggers/obj.panels, 以 obj.get('triggers')/obj.get('panels') 取代之

`tag:changed` 去除 new Obj() 之后的 render(). 对于 Switchable 来说, 不需要手工调用 render(), 直接 new 完就是初始状态.
不过如果用户再次调用 render(), 也没关系. render 中有判断是否已经 render 过, 也不会重复绑定事件和插入 DOM

## 0.9.15

`tag:improved` 升级 arale/widget@1.1.1

## 0.9.14

`tag:improved` 对 arale/widget 的依赖从 1.0.3 升级到 1.1.0 。

## 0.9.12

`tag:fixed` [#9](https://github.com/aralejs/switchable/issues/9) Carousel 上页滚屏的时候出现空白页。

`tag:fixed` [#10](https://github.com/aralejs/switchable/issues/10) className 可以通过配置取消。

`tag:improved` 依赖升级到 arale.widget@1.0.3 。



src/plugins/effects.js
var $ = require('jquery');

require('arale-easing');

var SCROLLX = 'scrollx';
var SCROLLY = 'scrolly';
var FADE = 'fade';


// 切换效果插件
module.exports = {
  attrs: {
    // 切换效果，可取 scrollx | scrolly | fade 或直接传入 effect function
    effect: 'none',
    easing: 'linear',
    duration: 500
  },

  isNeeded: function () {
    return this.get('effect') !== 'none';
  },

  install: function () {
    var panels = this.get('panels');

    // 注：
    // 1. 所有 panel 的尺寸应该相同
    //    最好指定第一个 panel 的 width 和 height
    //    因为 Safari 下，图片未加载时，读取的 offsetHeight 等值会不对
    // 2. 初始化 panels 样式
    //    这些特效需要将 panels 都显示出来
    // 3. 在 CSS 里，需要给 container 设定高宽和 overflow: hidden
    panels.show();
    var effect = this.get('effect');
    var step = this.get('step');

    var isFunction = $.isFunction(effect);

    // 初始化滚动效果
    if (!isFunction && effect.indexOf('scroll') === 0) {
      var content = this.content;
      var firstPanel = panels.eq(0);

      // 设置定位信息，为滚动效果做铺垫
      content.css('position', 'relative');

      // 注：content 的父级不一定是 container
      if (content.parent().css('position') === 'static') {
        content.parent().css('position', 'relative');
      }

      // 水平排列
      if (effect === SCROLLX) {
        panels.css('float', 'left');
        // 设置最大宽度，以保证有空间让 panels 水平排布
        // 35791197px 为 360 下 width 最大数值
        content.width('35791197px');
      }

      // 只有 scrollX, scrollY 需要设置 viewSize
      // 其他情况下不需要
      var viewSize = this.get('viewSize');
      if (!viewSize[0]) {
        viewSize[0] = firstPanel.outerWidth() * step;
        viewSize[1] = firstPanel.outerHeight() * step;
        this.set('viewSize', viewSize);
      }

      if (!viewSize[0]) {
        throw new Error('Please specify viewSize manually');
      }
    }
    // 初始化淡隐淡出效果
    else if (!isFunction && effect === FADE) {
      var activeIndex = this.get('activeIndex');
      var min = activeIndex * step;
      var max = min + step - 1;

      panels.each(function (i, panel) {
        var isActivePanel = i >= min && i <= max;
        $(panel).css({
          opacity: isActivePanel ? 1 : 0,
          position: 'absolute',
          zIndex: isActivePanel ? 9 : 1
        });
      });
    }

    // 覆盖 switchPanel 方法
    this._switchPanel = function (panelInfo) {
      var effect = this.get('effect');
      var fn = $.isFunction(effect) ? effect : Effects[effect];
      fn.call(this, panelInfo);
    };
  }
};


// 切换效果方法集
var Effects = {

  // 淡隐淡现效果
  fade: function (panelInfo) {
    // 简单起见，目前不支持 step > 1 的情景。若需要此效果时，可修改结构来达成。
    if (this.get('step') > 1) {
      throw new Error('Effect "fade" only supports step === 1');
    }

    var fromPanel = panelInfo.fromPanels.eq(0);
    var toPanel = panelInfo.toPanels.eq(0);

    if (this.anim) {
      // 立刻停止，以开始新的
      this.anim.stop(false, true);
    }

    // 首先显示下一张
    toPanel.css('opacity', 1);
    toPanel.show();

    if (panelInfo.fromIndex > -1) {
      var that = this;
      var duration = this.get('duration');
      var easing = this.get('easing');

      // 动画切换
      this.anim = fromPanel.animate({
        opacity: 0
      }, duration, easing, function () {
        that.anim = null; // free
        // 切换 z-index
        toPanel.css('zIndex', 9);
        fromPanel.css('zIndex', 1);
        fromPanel.css('display', 'none');
      });
    }
    // 初始情况下没有必要动画切换
    else {
      toPanel.css('zIndex', 9);
    }
  },

  // 水平/垂直滚动效果
  scroll: function (panelInfo) {
    var isX = this.get('effect') === SCROLLX;
    var diff = this.get('viewSize')[isX ? 0 : 1] * panelInfo.toIndex;

    var props = {};
    props[isX ? 'left' : 'top'] = -diff + 'px';

    if (this.anim) {
      this.anim.stop();
    }

    if (panelInfo.fromIndex > -1) {
      var that = this;
      var duration = this.get('duration');
      var easing = this.get('easing');

      this.anim = this.content.animate(props, duration, easing, function () {
        that.anim = null; // free
      });
    }
    else {
      this.content.css(props);
    }
  }
};

Effects[SCROLLY] = Effects.scroll;
Effects[SCROLLX] = Effects.scroll;
module.exports.Effects = Effects;